pkg/controller/certificaterequests/acme/acme_test.go

```diff
@@ -0,0 +1,211 @@
+	cmlisters "github.com/jetstack/cert-manager/pkg/client/listers/certmanager/v1alpha1"
+	testcr "github.com/jetstack/cert-manager/pkg/controller/certificaterequests/test"
+	testpkg "github.com/jetstack/cert-manager/pkg/controller/test"
+	"github.com/jetstack/cert-manager/test/unit/gen"
+	testlisters "github.com/jetstack/cert-manager/test/unit/listers"
+)
+
+func buildOrderMust(t *testing.T, cr *v1alpha1.CertificateRequest,
+	csr *x509.CertificateRequest) *v1alpha1.Order {
+	order, err := buildOrder(cr, csr)
+	if err != nil {
+		t.Errorf("failed to build order during testing: %s", err)
+		t.FailNow()
+	}
+
+	return order
+}
+
+func TestSign(t *testing.T) {
+	rsaPK := testcr.GenerateRSAPrivateKey(t)
+	csr, csrPEM := testcr.GenerateCSR(t, rsaPK)
+
+	testCR := gen.CertificateRequest("test-cr",
+		gen.SetCertificateRequestCSR(csrPEM),
+		gen.SetCertificateRequestIsCA(true),
+		gen.SetCertificateRequestDuration(&metav1.Duration{Duration: time.Hour * 24 * 60}),
+		gen.SetCertificateRequestIssuer(v1alpha1.ObjectReference{
+			Name:  "acme-issuer",
+			Group: certmanager.GroupName,
+			Kind:  "Issuer",
+		}),
+	)
+
+	_, rsaPEMCert := testcr.GenerateSelfSignedCertFromCR(t, testCR, rsaPK, time.Hour*24*60)
+	testOrder := buildOrderMust(t, testCR, csr)
+
+	tests := map[string]testT{
+		"a badly formed CSR should report failure": {
+			issuer: gen.Issuer("acme-issuer"),
+			certificateRequest: gen.CertificateRequestFrom(testCR,
+				gen.SetCertificateRequestCSR([]byte("a bad csr")),
+			),
+			builder: &testpkg.Builder{
+				KubeObjects:        []runtime.Object{},
+				CertManagerObjects: []runtime.Object{},
+				ExpectedEvents: []string{
+					"Warning ErrorParsingCSR Failed to decode CSR in spec: error decoding certificate request PEM block",
+				},
+				CheckFn: testcr.MustNoResponse,
+			},
+			expectedErr: false,
+		},
+
+		//TODO: Think of a creative way to get `buildOrder` to fail :thinking_face:
+
+		"if order doesn't exist then attempt to create one": {
+			issuer:             gen.Issuer("acme-issuer"),
+			certificateRequest: testCR,
+			builder: &testpkg.Builder{
+				CertManagerObjects: []runtime.Object{},
+				ExpectedActions: []testpkg.Action{
+					testpkg.NewAction(coretesting.NewCreateAction(
+						v1alpha1.SchemeGroupVersion.WithResource("orders"),
+						gen.DefaultTestNamespace,
+						testOrder,
+					)),
+				},
+				ExpectedEvents: []string{
+					`Normal OrderPending Waiting on certificate issuance from order default-unit-test-ns/test-cr-3958469914: order is currently pending: ""`,
+				},
+				CheckFn: testcr.MustNoResponse,
+			},
+			expectedErr: false,
+		},
+
+		"if we fail to get the order resource then we should report pending": {
+			issuer:             gen.Issuer("acme-issuer"),
+			certificateRequest: testCR,
+			builder: &testpkg.Builder{
+				ExpectedEvents: []string{
+					`Normal ErrorGettingOrder Failed to get order resource default-unit-test-ns/test-cr-3958469914: this is a network error`,
```

Comment:
Undecided whether network errors etc. should be bubbled up the events, as it could in turn place even more burden on the network. But let's think about that later 🙄